### PR TITLE
Make the spawned process cancelable

### LINKED ISSUE #113: Make it cancelable
Using: https://github.com/sindresorhus/p-cancelable

It's possible to call `cp.kill()`, but would be nicer if it was cancelable.

### PR DESCRIPTION
**Pull request by ammarbinfaisal**
see if this is okay.
I was wondering whether there should we a check if oncancel property in options object passed by someone is actually a function. is this required?

Fixes #113

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -251,6 +252,8 @@ module.exports = (command, args, options) => {
 				resolve({error});
 			});
 		}
+
+		oncancel(parsed.options.oncancel);
```
Comment (sindresorhus) [inline index.js]: This should kill the process, not execute a function from an option.

File: index.js [hunk 2]
```diff
@@ -234,7 +235,7 @@ module.exports = (command, args, options) => {
 		}, parsed.options.timeout);
 	}
 
-	const processDone = new Promise(resolve => {
+	const processDone = new PCancelable((resolve, reject, oncancel) => {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	const processDone = new PCancelable((resolve, reject, onCancel) => {
```

**Comment by sindresorhus** [comment]
You need to add tests and docs too.

File: index.js (hunk 1 above)
Comment (ammarbinfaisal) [reply index.js]: done

**Comment by ammarbinfaisal** [comment]
There is CancelError being thrown. In test.js and readme, I've passed an empty function `() => {}` to spawned.catch to ignore it. should I use a try catch block inside the main code to do this?

**Comment by sindresorhus** [comment]
No, you're missing the point of CancelError. Please read the `p-cancelable` readme and understand how it works.

File: index.js [hunk 3]
```diff
@@ -311,6 +316,7 @@ module.exports = (command, args, options) => {
 	// eslint-disable-next-line promise/prefer-await-to-then
 	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
 	spawned.catch = onRejected => handlePromise().catch(onRejected);
+	spawned.cancel = reason => processDone.cancel(reason);
```
Comment (ehmicky) [inline index.js]: Maybe you should provide a default `reason` if none is passed, like `The command was canceled`

**Comment by ehmicky** [comment]
I think it would be nice to add properties similar to how `timedOut`, `failed` and `killed` and currently handled:
  - add `result.canceled`. Always defined (`true` or `false`) whether on success or failure. Exception: not defined with `sync()` since synchronous calls cannot be canceled.
  - enhance `result.message` (`getErrorPrefix()` function) to add `was canceled: ${reason}`. `: ${reason}` should not be appended in it's default `reason`.

File: readme.md [hunk 4]
```diff
@@ -51,6 +51,13 @@ const execa = require('execa');
 	const {stdout} = await execa.shell('echo unicorns');
 	//=> 'unicorns'
 
+	// Cancelling a spawned process
+	const spawned = execa("echo unicorns");
+	spawned.catch( error => {
+		// This would be equal to the reason provided for canceling
```
Comment (ehmicky) [inline readme.md]: This comment should probably be removed because `error.message` is actually changed by `execa` so `error.message` would not be exactly `"reason for canceling"`.

File: readme.md [hunk 5]
```diff
@@ -51,6 +51,13 @@ const execa = require('execa');
 	const {stdout} = await execa.shell('echo unicorns');
 	//=> 'unicorns'
 
+	// Cancelling a spawned process
+	const spawned = execa("echo unicorns");
+	spawned.catch( error => {
```
Comment (ehmicky) [inline readme.md]: Minor coding style: no space after parenthesis here.

File: readme.md [hunk 6]
```diff
@@ -51,6 +51,13 @@ const execa = require('execa');
...
+		// This would be equal to the reason provided for canceling
+		console.log(error.message);
+	}); // To handle the CancelError
```
Comment (ehmicky) [inline readme.md]: I think this comment will be a little obscure to readers, so could be left out.

File: readme.md [hunk 7]
```diff
@@ -51,6 +51,13 @@ const execa = require('execa');
...
+	spawned.cancel("reason for canceling");
```
Comment (ehmicky) [inline readme.md]: To simplify the example I would personally call `cancel()` without arguments here, but it's up to you if you think it's better.

File: test.js [hunk 8]
```diff
@@ -569,3 +569,16 @@ if (Promise.prototype.finally) {
 		t.is(result.message, 'called');
 	});
 }
+
+test('the spawned process is cancelable', t => {
+	const spawned = execa('ls');
```
Comment (ehmicky) [inline test.js]: `ls` is not available on Windows. You should use one of the test fixtures e.g. `node fixtures/forever`

File: test.js [hunk 9]
```diff
@@ -569,3 +569,16 @@ if (Promise.prototype.finally) {
...
+	spawned.catch(error => {
+		if (error.message === 'reason' && error.isCanceled) {
+			return;
+		}
+
+		throw error;
+	}); // To handle the CancelError
+	spawned.cancel('reason');
+	t.true(spawned.killed);
```
Comment (ehmicky) [inline test.js]: I think `canceled` should be tested here instead of `killed`. See my other comment about adding that property.

File: test.js [hunk 10]
```diff
@@ -569,3 +569,16 @@ if (Promise.prototype.finally) {
 		t.is(result.message, 'called');
 	});
 }
+
+test('the spawned process is cancelable', t => {
+	const spawned = execa('ls');
+	spawned.catch(error => {
```
Comment (ehmicky) [inline test.js]: I think you should instead use `await t.throwsAsync(spawned, ...)`

File: test.js [hunk 11]
```diff
@@ -569,3 +569,16 @@ if (Promise.prototype.finally) {
...
+		if (error.message === 'reason' && error.isCanceled) {
+			return;
+		}
+
+		throw error;
+	}); // To handle the CancelError
+	spawned.cancel('reason');
```
Comment (ehmicky) [inline test.js]: I think this test should be split into several tests each testing different parts of the PR:
  - test that after `cancel()` `result.canceled` is `true`
  - test that without `cancel()` `result.canceled` is `false`
  - test that after `cancel()` `promise is rejected with the right error
  - test that the message passed as a `reason` appears somewhere inside `result.message`

File: readme.md (hunk 7 above)
Comment (ammarbinfaisal) [reply readme.md]: what you are inferring seems better

File: test.js (hunk 11 above)
Comment (ammarbinfaisal) [reply test.js]: I'd have to explore ava's docs, I never written tests before :sweat_smile:

Comment (ehmicky) [reply test.js]: If you've never written tests before, I think this might a great opportunity for you here. [Ava](https://github.com/avajs/ava) is a great project and it's beginner-friendly.

For me the main thing is:
  - keeping tests simple: each test does one thing, then check if the expected result is correct.
  - splitting tests into small tests that each test one thing.
  - keeping tests independent from each other. Ava runs tests in parallel by default so it enforces avoiding using any global state.
  - using the tools the test runner provides (here Ava), e.g. `t.throwsAsync()` assertions, etc.
  - making sure you're covering all the features you're adding and things that could break.
  - remembering target platform: Windows/Mac/Linux, Node.js `>= 8`

Please ask any questions you might have while learning Ava and writing those tests, we're here to help :)

Comment (ammarbinfaisal) [reply test.js]: thanks a lot! actually I am a bit busy right now. I'll do all this tonight (5-6 hours from now)

Comment (ehmicky) [reply test.js]: There is no rush :+1:

**Comment by ammarbinfaisal** [comment]
> I think it would be nice to add properties similar to how timedOut, failed and killed and currently handled:

add result.canceled. Always defined (true or false) whether on success or failure. Exception: not defined with sync() since synchronous calls cannot be canceled.
enhance result.message (getErrorPrefix() function) to add was canceled: ${reason}. : ${reason} should not be appended in it's default reason.

@ehmicky, I can't figure out how to do this. On cancelling p-cancelable itself throws CancelError and this is not notified in the handlePromise function - I mean in the if block which checks whether there is an error and and creates a proper execa error using makeError.

File: test.js [hunk 12]
```diff
@@ -569,3 +570,30 @@ if (Promise.prototype.finally) {
...
+test('invoking cancel method throws CancelError', async t => {
+	const spawned = execa('ls');
+	spawned.cancel();
+	await t.throwsAsync(spawned, CancelError);
+});
+
+test('default CancelError message is "command was canceled"', async t => {
+	const spawned = execa('ls');
+	spawned.cancel();
+	const error = await t.throwsAsync(spawned);
```
Comment (ehmicky) [inline test.js]: You can do those two lines in a single one: `await t.throwsAsync(spawned, { message: 'command was canceled' })`

Same thing for the error tests.

See https://github.com/avajs/ava/blob/master/docs/03-assertions.md#throwsasyncthrower-expected-message for more information.

**Comment by ehmicky** [comment]
Something like this might work if I understand your problem correctly:

```js
	const processDone = new PCancelable((resolve, reject, onCancel) => {
        let canceled = false;
		spawned.on('exit', (code, signal) => {
			cleanup();
			resolve({code, signal, canceled});
		});

		/* more code */

		onCancel(() => {
            canceled = true;
			spawned.kill();
		});
	});
```

**Comment by ammarbinfaisal** [comment]
> Something like this might work if I understand your problem correctly:

@ehmicky No, it won't. I also thought of taking this variable in handlePromise function. As soon as we cancel the promise p-cancelable itself throws the error as you can see [here](https://github.com/sindresorhus/p-cancelable/blob/master/index.js).

**Comment by ehmicky** [comment]
This might work if you put the `canceled` variable one level up:

```js
let canceled = false;
const processDone = new PCancelable((resolve, reject, onCancel) => {
  /* more code */
  onCancel(() => {
    canceled = true;
    spawned.kill();
  });
});

const handlePromise = () => pFinally(Promise.all([
		processDone,
		getStream(spawned, 'stdout', {encoding, buffer, maxBuffer}),
		getStream(spawned, 'stderr', {encoding, buffer, maxBuffer})
	]).then(results => { // eslint-disable-line promise/prefer-await-to-then
		const result = results[0];
		result.stdout = results[1];
		result.stderr = results[2];

		if (result.error || result.code !== 0 || result.signal !== null || canceled) {
			const error = makeError(result, {
				joinedCommand,
				parsed,
				timedOut,
                canceled
			});

			// TODO: missing some timeout logic for killed
			// https://github.com/nodejs/node/blob/master/lib/child_process.js#L203
			// error.killed = spawned.killed || killed;
			error.killed = error.killed || spawned.killed;

			if (!parsed.options.reject) {
				return error;
			}

			throw error;
		}

		return {
			stdout: handleOutput(parsed.options, result.stdout),
			stderr: handleOutput(parsed.options, result.stderr),
			code: 0,
			failed: false,
			killed: false,
			signal: null,
			cmd: joinedCommand,
			timedOut: false,
			canceled: false,
		};
	}), destroy);
```

Notice the `|| canceled` condition: if the promise is rejected because of cancellation, exit code might still be `0` with no errors nor signals (in case child process killing happens on the next tick), but we want to make sure cancellation is taken into account.

**Comment by ehmicky** [comment]
Ok I just realized from my above comment that the `processDone` promise will be rejected with a `CancelError`, which won't work with the rest of the code which assumes otherwise.

However when we cancel the promise, the child process is killed. This will lead to the `exit` event to be called, which will resolve the promise. Also we do want to wait on the child process to be killed before returning the result. 

So I suggest two solutions. First using [`onCancel.shouldReject = false`](https://github.com/sindresorhus/p-cancelable/blob/master/readme.md#new-pcancelableexecutor).

Second not using `p-cancelable` and simply doing:

```js
	const processDone = new Promise((resolve, reject) => {
		spawned.on('exit', (code, signal) => {
			cleanup();
			resolve({code, signal});
		});

		spawned.on('error', error => {
			cleanup();
			resolve({error});
		});

		if (spawned.stdin) {
			spawned.stdin.on('error', error => {
				cleanup();
				resolve({error});
			});
		}
	});

    let canceled = false;
    processDone.cancel = function() {
      if (canceled) { return; }

      canceled = true;
      spawned.kill();
    }
```

**Comment by ammarbinfaisal** [comment]
> This might work if you put the canceled variable one level up

I actually tried doing all this.

`
	// eslint-disable-next-line promise/prefer-await-to-then
	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
	spawned.catch = onRejected => handlePromise().catch(onRejected);
	spawned.cancel = (reason = 'command was canceled') => processDone.cancel(reason);  // the problem is with this line

`

Since, calling spawned.cancel class processDone.cancel, CancelError is thrown and proccess simply terminates. The callback would be executed inside onCancel function and canceled would become true however that if block won't be  executed so we can't make a proper execa error and do the rest of thing you are thinking of. atleast, i can't figue out a way. We can prevent cancelError from being thrown([https://github.com/sindresorhus/p-cancelable#new-pcancelableexecutor](see this)) and then make a new error on our own if canceled is `true`. Will that be okay?

**Comment by ammarbinfaisal** [comment]
@ehmicky  I can't understand why is this happening but if I use `noop`, `node forever` instead of ls in the latest added test, it fails

```
  1 test failed

  test › calling cancel method makes result.cancel true


  Rejected promise returned by test. Reason:

  Error {
    cmd: 'node fixtures',
    code: 2,
    errno: 'ENOENT',
    failed: true,
    killed: false,
    path: 'node fixtures',
    signal: null,
    spawnargs: [],
    stderr: '',
    stdout: '',
    syscall: 'spawn node fixtures',
    timedOut: false,
    message: 'Command failed with exit code 2 (ENOENT): spawn node fixtures ENOENT',
  }

```

**Comment by ehmicky** [comment]
Could you post the code of that test?

**Comment by ammarbinfaisal** [comment]
this is the failing test

```
test('calling cancel method makes result.cancel true', async t => {
	const spawned = execa('ls');
	spawned.cancel();
	const result = await spawned;
	t.true(result.canceled);
});
```

I think this depends on the time taken by the process like even with ls it failed one time on my PC and ran the next time
Screenshot from 2019-03-13 19-49-10

**Comment by ammarbinfaisal** [comment]
I think using p-cancelable and allowing CancelError to be thrown is the best think which can be done; probably @sindresorhus asked to use p-cancelable because of that and why add a canceled property in result when the person using this can just check for CancelError(i mean use catch) :thinking:

**Comment by ehmicky** [comment]
You're showing a different test that the initial failure. The initial failure you mentioned was about `Command failed with exit code 2 (ENOENT): spawn node fixtures ENOENT`. The one you are now posting uses a different command and is a separate problem. Let's try not to fork too many different discussions, otherwise it's hard to keep track.

What you are trying is also dependent on how you will implement this feature. If there is a timing issue there, the issue is not the test but the implementation.

About letting the `CancelError` exception propagate: the issue there is that:
  1) while `process.kill()` has been called, the child process might not be called right away (synchronously) by Node.js, i.e. `execa()` might resolve while the child process is still being killed. This is probably more proper to only resolve `execa()` once the child process is killed.
  2) because of 1), the result will be missing the termination `signal` and the `code`
  3) it adds a new value to which `processDone()` can resolve, which leads to more polymorphic code, i.e. more complexity.

Either of the two solutions above would solve those issues (or any third solution you might think of).

**Comment by ammarbinfaisal** [comment]
Sorry, I didn't see that was ENOENT.
I think I have done what you asked for or am I missing something?

File: index.js [hunk 13]
```diff
@@ -311,6 +322,7 @@ module.exports = (command, args, options) => {
 	// eslint-disable-next-line promise/prefer-await-to-then
 	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
 	spawned.catch = onRejected => handlePromise().catch(onRejected);
+	spawned.cancel = () => processDone.cancel();
```
Comment (ehmicky) [inline index.js]: This could be merged with `processDone.cancel()`. I.e. if `processDone.cancel = function {` was `spawned.cancel = function () {` instead, then this line of code could be removed.

Comment (ammarbinfaisal) [reply index.js]: You mean it should be like this 
```
spawned.cancel = processDone.cancel;
```

**Comment by ehmicky** [comment]
Ah ok indeed the code was changed. I thought otherwise because of your comment:

> I think using p-cancelable and allowing CancelError to be thrown is the best think which can be done

For the test, let's break it down into two different problems:
  a) You had an `ENOENT` error. Please let me know if the fix I suggested above is working.
  b) You had a timing / race condition issue. I think the new implementation might have fixed that problem.

File: index.js (hunk 13 above)
Comment (ehmicky) [reply index.js]: No:
  - removing the `spawned.cancel = () => processDone.cancel()'` line
  - replacing `processDone.cancel = function {` by `spawned.cancel = function () {`

File: package.json [hunk 14]
```diff
@@ -41,6 +41,7 @@
 		"get-stream": "^4.0.0",
 		"is-stream": "^1.1.0",
 		"npm-run-path": "^2.0.0",
+		"p-cancelable": "^1.1.0",
```
Comment (ehmicky) [inline package.json]: This can be removed if it's not used.

**Comment by ammarbinfaisal** [comment]
> You had an ENOENT error. Please let me know if the fix I suggested above is working.

that ENOENT problem isn't there anymore using `noop`; still, this happens

```
1 test failed

  test › calling cancel method makes result.cancel true

  /home/ammar/osource/execa/index.js:5

   4: const childProcess = require('child_process');           
   5: const crossSpawn = require('cross-spawn');               
   6: const stripFinalNewline = require('strip-final-newline');

  Rejected promise returned by test. Reason:

  Error {
    cmd: 'noop',
    code: undefined,
    failed: true,
    killed: true,
    signal: 'SIGTERM',
    stderr: '',
    stdout: '',
    timedOut: false,
    message: 'Command was killed with SIGTERM: noop',
  }

  makeError (index.js:5:1171)
  Promise.all.then.results (index.js:7:376)
```

<br>

**One thing more**

```
const execa = require("../execa");
const spawned = execa("ls");

spawned.cancel();
spawned.then(result => {
    console.log(result.canceled, spawned.killed)
}).catch(error => console.log(error));
```
this logs
`true true`
but if I replace `ls` with `tree` then
```
{ Error: Command was killed with SIGTERM: tree
    at makeError (/home/ammar/osource/execa/index.js:133:11)
    at Promise.all.then.results (/home/ammar/osource/execa/index.js:278:18)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  code: undefined,
  stdout: '',
  stderr: '',
  failed: true,
  signal: 'SIGTERM',
  cmd: 'tree',
  timedOut: false,
  killed: true }
```

File: index.js [hunk 15]
```diff
@@ -253,6 +254,15 @@ module.exports = (command, args, options) => {
 		}
 	});
 
+	processDone.cancel = function () {
+		if (canceled) {
+			return;
+		}
+
+		canceled = true;
+		spawned.kill();
```
Comment (ehmicky) [inline index.js]: Probably this should actually be:

```js
if (spawned.kill()) {
  canceled = true;
}
```

Otherwise trying to `cancel()` an already killed process will set it `canceled` although `cancel()` did nothing.

See Node source code: https://github.com/nodejs/node/blob/master/lib/internal/child_process.js#L435

**Comment by ehmicky** [comment]
Both code samples seem to be wrong:
  - in the first one, `spawned` should be rejected (since it was canceled), i.e. `catch()` callback should be called, not `then()` callback
  - the second is almost correct except the return value should include `canceled: true`

File: index.js [hunk 16]
```diff
@@ -322,7 +318,12 @@ module.exports = (command, args, options) => {
 	// eslint-disable-next-line promise/prefer-await-to-then
 	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
 	spawned.catch = onRejected => handlePromise().catch(onRejected);
-	spawned.cancel = () => processDone.cancel();
+	spawned.cancel = function () {
+		if (!spawned.killed) {
```
Comment (ammarbinfaisal) [inline index.js]: I did `!spawned.killed` instead of `spawn.kill()` because doing that would give another error(prcess was killed with SIGTERM) and not the one with message "Command was canceled" and property canceled equal to `true`(I think this is the required thing)

Comment (ehmicky) [reply index.js]: This sounds good but I think you should still check the return value of `spawned.kill()`. For example if the child process exited successfully (not killed) `spawned.kill()` will return `false`. In that case it was not canceled, and `canceled` should not be `true`.

File: test.js [hunk 17]
```diff
@@ -586,3 +586,42 @@ if (Promise.prototype.finally) {
...
+	});
+	t.regex(error.message, /Command\swas\scanceled/);
+});
+
+test('error.canceled is true when cancel method is used', async t => {
+	const spawned = execa('noop');
+	spawned.cancel();
+	const error = await t.throwsAsync(async () => {
+		await spawned;
+	});
```
Comment (ehmicky) [inline test.js]: You can directly use `t.throwsAsync(spawned)` (same thing for the other tests).

File: test.js [hunk 18]
```diff
@@ -586,3 +586,42 @@ if (Promise.prototype.finally) {
...
+	t.false(result.canceled);
+});
+
+test('calling cancel method throws an error with message "Command was canceled"', async t => {
+	const spawned = execa('noop');
+	spawned.cancel();
+	const error = await t.throwsAsync(async () => {
+		await spawned;
+	});
+	t.regex(error.message, /Command\swas\scanceled/);
```
Comment (ehmicky) [inline test.js]: You can directly use `t.throwsAsync(spawned, { message: /Command was canceled/ })`

File: test.js [hunk 19]
```diff
@@ -586,3 +586,42 @@ if (Promise.prototype.finally) {
...
+});
```
Comment (ehmicky) [inline test.js]: I think the regExp can use spaces instead of `\s` because we know for sure those will just be single spaces.

File: test.js [hunk 20]
```diff
@@ -586,3 +586,42 @@ if (Promise.prototype.finally) {
...
+
+test('cancel method kills the spawned process', t => {
+	const spawned = execa('node');
+	spawned.cancel();
+	t.true(spawned.killed);
+});
+
+test('result.canceled is false when spawned.cancel isn\'t called', async t => {
+	const spawned = execa('noop');
+	const result = await spawned;
```
Comment (ehmicky) [inline test.js]: You can directly do `const result = await execa('noop')`

File: readme.md (hunk 6 above)
Comment (ehmicky) [reply readme.md]: Actually both comments can probably be left out above.

File: test.js [hunk 21]
```diff
@@ -586,3 +586,42 @@ if (Promise.prototype.finally) {
...
+test('error.canceled is true when cancel method is used', async t => {
+	const spawned = execa('noop');
+	spawned.cancel();
+	const error = await t.throwsAsync(async () => {
+		await spawned;
+	});
+	t.true(error.canceled);
+});
+
+test('error.canceled is false when kill method is used', async t => {
```
Comment (ehmicky) [inline test.js]: Except for comments above, tests look good now!

Just three extra tests that you might want to add: 
  - calling `spawned.cancel()` twice should have same behavior as calling it once.
  - calling `spawned.cancel()` on a child process that has already successfully exited
  - calling `spawned.cancel()` on a child process that has already been killed

File: index.js [hunk 22]
```diff
@@ -358,7 +358,9 @@ module.exports = (command, args, options) => {
 	spawned.cancel = function () {
 		if (!spawned.killed) {
 			canceled = true;
-			spawned.kill();
+			if (!spawned.kill()) {
+				canceled = false;
+			}
 		}
```
Comment (ehmicky) [inline index.js]: If you call twice `cancel()`, the second time `canceled` will be set back to `false`.

This could be fixed by doing instead:

```js
spawned.cancel = function() {
  if (spawned.killed) { return; }

  if (spawned.kill()) {
    canceled = true;
  }
};
```

File: test.js [hunk 23]
```diff
@@ -586,3 +586,56 @@ if (Promise.prototype.finally) {
...
+test('result.canceled is false when spawned.cancel isn\'t called', async t => {
+	const result = await execa('noop');
+	t.false(result.canceled);
+});
+
+test('calling cancel method throws an error with message "Command was canceled"', async t => {
+	const spawned = execa('noop');
+	spawned.cancel();
+	const error = await t.throwsAsync(spawned);
+	t.regex(error.message, /Command was canceled/);
```
Comment (ehmicky) [inline test.js]: You can directly use t.throwsAsync(spawned, { message: /Command was canceled/ })

Comment (ammarbinfaisal) [reply test.js]: yeah, sorry for forgetting this

File: index.js [hunk 24]
```diff
@@ -356,7 +356,7 @@ module.exports = (command, args, options) => {
 	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
 	spawned.catch = onRejected => handlePromise().catch(onRejected);
 	spawned.cancel = function () {
-		if (!spawned.killed) {
+		if (!spawned.killed && spawned.exitCode !== 0) {
```
Comment (ehmicky) [inline index.js]: In which situation can `spawned` be `killed` but with an `exitCode` of `0`?

Comment (ammarbinfaisal) [reply index.js]: oh sorry! I should check for something else...but that won't work

Comment (ammarbinfaisal) [reply index.js]: I am not wrong then exitcode is 0 when the process was successful. this would just check whether the proccess hasn't successfully exited and can be no longer killed

Comment (ehmicky) [reply index.js]: As far as I understand the only thing we need to check is whether the process has already terminated. It should not matter whether it was successful or not, and whether it was done through a signal (`killed`) or not. 

This is already given by Node.js with the return value of `spawned.kill()` so wouldn't this work instead?

```js
if (canceled) { return }

if (process.kill()) {
  canceled = true;
}
```

File: readme.md (hunk 6 above)
Comment (ammarbinfaisal) [reply readme.md]: done

File: index.js (hunk 24 above)
Comment (ammarbinfaisal) [reply index.js]: This would work but then what would be the use of `canceled` at all. We won't be able to use it. On killing the process, processDone promise would get resolved

```
const processDone = new Promise(resolve => {
                // some code
		spawned.on('error', error => {
			cleanup();
			resolve({error});  // because of this
		});
                // more code
	});
```

and then in handlePromise function this condition check would pass with canceled still being `false`.

```
if (result.error || result.code !== 0 || result.signal !== null || canceled) {
			const error = makeError(result, {
				joinedCommand,
				parsed,
				timedOut,
				canceled
			});
```
 Thus, the error made by makeError won't be as desired - having canceled == true and a proper message

Comment (ammarbinfaisal) [reply index.js]: all what is mean is that we can;t execute `kill()` before making `canceled` `true`

Comment (ehmicky) [reply index.js]: Isn't `handlePromise` only resolved on the next tick?  I.e. the order would be: 
  1) `spawned.kill()` is fired
  2) `canceled = true` is executed
  3) `handlePromise` callback is executed

Correct me if I'm wrong.

Comment (ammarbinfaisal) [reply index.js]: oh! this does work. not sure why but I tried this in morning(kill before making canceled true) and this threw that error.

**Comment by ammarbinfaisal** [comment]
sorry for troubling you so much over this PR.

File: index.js (hunk 24 above)
Comment (ehmicky) [reply index.js]: You actually made me double check, I was not sure either whether `exit` would be fired on the next tick or not.

**Comment by ehmicky** [comment]
You are helping us add a new feature, there's nothing to be sorry about! 

Code review is part of the process and I enjoy it, so no worries.

Hopefully this is a good experience for you as well, in case you wanted to contribute again after this PR :)

File: readme.md [hunk 25]
```diff
@@ -53,8 +53,14 @@ const execa = require('execa');
 	//=> 'unicorns'
 
 	// Cancelling a spawned process
-	const spawned = execa("echo unicorns");
+	const spawned = execa("node");
 	spawned.cancel();
```
Comment (ehmicky) [inline readme.md]: Should we put this in a `setTimeout(() => { spawned.cancel() }, 1000)` block? So that it's clear the process can be canceled after being `await`ed?

File: readme.md [hunk 26]
```diff
@@ -149,6 +155,13 @@ Execute a command synchronously through the system shell.
 
 Returns the same result object as [`child_process.spawnSync`](https://nodejs.org/api/child_process.html#child_process_child_process_spawnsync_command_args_options).
 
+### spawned.cancel()
+
+Cancel a process spawned using execa. Calling this method kills it.
+
+Throws an error with `error.canceled` equal to `true`, provided that the process gets canceled.
+Process would not get canceled if it has already exited or has been killed by `spawned.kill()`.
```
Comment (ehmicky) [inline readme.md]: I think you can just stop at `already exited` since if it has been killed, then it has exited.

**Comment by ehmicky** [comment]
@sindresorhus the PR looks good to me, but awaiting on your own code review before merging.

File: index.js [hunk 27]
```diff
@@ -334,7 +341,8 @@ module.exports = (command, args, options) => {
 			killed: false,
 			signal: null,
 			cmd: joinedCommand,
-			timedOut: false
+			timedOut: false,
+			canceled: false
```
Comment (sindresorhus) [inline index.js]: I would prefer to name it `isCanceled`. I realize it's inconsistent with `killed`, but I want to rename that one at some point too. Use the `isCanceled` naming for all the `canceled` variables.

File: index.js [hunk 28]
```diff
@@ -347,6 +355,15 @@ module.exports = (command, args, options) => {
 	// eslint-disable-next-line promise/prefer-await-to-then
 	spawned.then = (onFulfilled, onRejected) => handlePromise().then(onFulfilled, onRejected);
 	spawned.catch = onRejected => handlePromise().catch(onRejected);
+	spawned.cancel = function () {
```
Comment (sindresorhus) [inline index.js]: Use arrow function

File: readme.md [hunk 29]
```diff
@@ -52,6 +52,15 @@ const execa = require('execa');
 	const {stdout} = await execa.shell('echo unicorns');
 	//=> 'unicorns'
 
+	// Cancelling a spawned process
+	const spawned = execa("node");
```
Comment (sindresorhus) [inline readme.md]: Single-quotes

Comment (sindresorhus) [reply readme.md]: I think `process` would be a better name than `spawned`.

File: readme.md [hunk 30]
```diff
@@ -146,6 +155,13 @@ Execute a command synchronously through the system shell.
 
 Returns the same result object as [`child_process.spawnSync`](https://nodejs.org/api/child_process.html#child_process_child_process_spawnsync_command_args_options).
 
+### spawned.cancel()
```
Comment (sindresorhus) [inline readme.md]: It's not clear enough what `spawned` is. And it's weird to place the docs here. I think it would be better to document it in the `execa()` section.

File: test.js [hunk 31]
```diff
@@ -586,3 +586,57 @@ if (Promise.prototype.finally) {
...
+	t.true(spawned.killed);
+});
+
+test('result.canceled is false when spawned.cancel isn\'t called', async t => {
+	const result = await execa('noop');
+	t.false(result.canceled);
+});
+
+test('calling cancel method throws an error with message "Command was canceled"', async t => {
+	const spawned = execa('noop');
```
Comment (sindresorhus) [inline test.js]:
```suggestion
	const process = execa('noop');
```

File: readme.md (hunk 29 above)
Comment (ehmicky) [reply readme.md]: Would this confuse readers about the global variable named `process`? Child processes returned by `spawn()` and global `process` have different methods/properties (although some are shared).

Comment (sindresorhus) [reply readme.md]: Ah yeah, `subprocess` is better.

File: readme.md [hunk 32]
```diff
@@ -121,6 +121,9 @@ Returns a [`child_process` instance](https://nodejs.org/api/child_process.html#c
 
 It exposes an additional `.all` stream, with `stdout` and `stderr` interleaved.
 
+It can be canceled with `.cancel` method which throws an error with `error.canceled` equal to `true`, provided that the process gets canceled.
```
Comment (ehmicky) [inline readme.md]: `canceled` -> `isCanceled()`

File: index.js [hunk 33]
```diff
@@ -152,7 +152,7 @@ function makeError(result, options) {
 		error = new Error(message);
 	}
 
-	const prefix = getErrorPrefix({timedOut, timeout, signal, exitCodeName, exitCode});
+	const prefix = getErrorPrefix({timedOut, timeout, signal, exitCode, exitCodeName, isCanceled});
```
Comment (sindresorhus) [inline index.js]: Don't do unrelated changes

**Comment by sindresorhus** [comment]
Thank you for contributing, @ammarbinfaisal :)

File: index.js (hunk 33 above)
Comment (ammarbinfaisal) [reply index.js]: I might have done this while fixing some merge conflict; that's the only time I touched those two variables `exitCode` and `exitCodeName`
